### PR TITLE
T2D-177 Oauth using EGA AAI and custom Role Based authority

### PR DESCRIPTION
**Pull request by selvaebi**

### CODE REVIEW

File: src/main/java/uk/ac/ebi/ega/accession/user/rest/AccessioningUserRestController.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import uk.ac.ebi.ega.accession.user.AccessioningUserRepository;
+
+@RestController
+@RequestMapping(value = "/users")
+public class AccessioningUserRestController {
+
+    @Autowired
+    private AccessioningUserRepository accessioningUserRepository;
+
+    @RequestMapping(path = "addOrUpdate", method = {RequestMethod.POST},
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/ega/accession/user/rest/AccessioningUserRestController.java]: RESTful APIs are based on the HTTP method only, so we shouldn't extend the URL with this `addOrUpdate`.

File: src/main/java/uk/ac/ebi/ega/accession/configuration/security/DisableSecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.oauth2.config.annotation.web.configuration.EnableResourceServer;
+import org.springframework.security.oauth2.config.annotation.web.configuration.ResourceServerConfigurerAdapter;
+
+@ConditionalOnProperty(value = "security.enabled", havingValue = "false")
+@Configuration
+@EnableResourceServer
+public class DisableSecurityConfig extends ResourceServerConfigurerAdapter {
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/ega/accession/configuration/security/DisableSecurityConfig.java]: What about using the property described [here](https://stackoverflow.com/a/34150562)?

Comment (selvaebi) [reply src/main/java/uk/ac/ebi/ega/accession/configuration/security/DisableSecurityConfig.java]: it is kind of deprecated I guess, that is used this config. https://spring.io/blog/2017/09/15/security-changes-in-spring-boot-2-0-m4

Comment (jorizci) [reply src/main/java/uk/ac/ebi/ega/accession/configuration/security/DisableSecurityConfig.java]: It is deprecated and also Oauth has gone a massive revamp, so, unfortunately in the future we will need to do some adjustments, as right now moving to that revamp would require us to up all the versions.

File: src/main/java/uk/ac/ebi/ega/accession/configuration/security/CustomUserAuthenticationConverter.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        return response;
+    }
+
+    public Authentication extractAuthentication(Map<String, ?> map) {
+        if (map.containsKey(USER_ID)) {
+            Object principal = map.get(USER_ID);
+            Collection<? extends GrantedAuthority> authorities = getAuthorities(map);
+            return new UsernamePasswordAuthenticationToken(principal, "N/A", authorities);
+        }
+        return null;
```
Comment (jorizci) [inline src/main/java/uk/ac/ebi/ega/accession/configuration/security/CustomUserAuthenticationConverter.java]: As discussed offline we need to double check what will be the behaviour with no user or user that does not exist on the database

File: src/main/java/uk/ac/ebi/ega/accession/configuration/security/CustomUserAuthenticationConverter.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * This class provides custom authorities for the EGA AAI authenticated user.
+ */
+public class CustomUserAuthenticationConverter implements UserAuthenticationConverter {
```
Comment (jorizci) [inline src/main/java/uk/ac/ebi/ega/accession/configuration/security/CustomUserAuthenticationConverter.java]: After giving some thought, this is not the proper way of handle this. The converter is just to pass the map to the authorization token it is not the proper way of getting this information (altough you can do it as you've done). The proper way would be to have configured correctly this: https://docs.spring.io/spring-security-oauth2-boot/docs/current/api/org/springframework/boot/autoconfigure/security/oauth2/resource/AuthoritiesExtractor.html

File: src/main/java/uk/ac/ebi/ega/accession/configuration/SwaggerConfiguration.java [hunk 5]
```diff
@@ -64,6 +68,12 @@
     @Autowired
     private SwaggerApiInfoProperties swaggerApiInfoProperties;
 
+    private SecurityReference securityReference = SecurityReference.builder()
+            .reference("Authorization").scopes(new AuthorizationScope[0]).build();
+
+    private SecurityContext securityContext = SecurityContext.builder()
```
Comment (jorizci) [inline src/main/java/uk/ac/ebi/ega/accession/configuration/SwaggerConfiguration.java]: Either do these initialization as beans or use them in the code. Avoid as much as possible complete initialization with this idiom.

**Pull request merged**